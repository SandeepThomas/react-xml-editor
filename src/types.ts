export interface MenuItemSpec {
    action: (xml: Xml, id: string[]) => Xml | Promise<Xml>;
    caption: string;
    hideIf?: (a: any) => boolean;
}

export type Asker = (options: AskerOptions) => React.ReactNode;

export interface AskPicklistCaptionMenuItem {
    caption: string;
    value: string
};

export type AskPicklistMenuItem = AskPicklistCaptionMenuItem | string;

export type AskerParameter = AskPicklistMenuItem[];
export interface AttributeSpec {
    asker?: Asker;
    menu?: MenuItemSpec[];
}

export interface ElementsSpec {
    asker?: Asker;
    attributes?: {[key: string]: AttributeSpec};
    menu?: MenuItemSpec[];
}

export interface DocSpec {
    elements?: {[key: string]: ElementsSpec};
}

export interface Xml {
    [key: string]: Element;
}

export enum BubbleType {
    ASKER,
    MENU,
}

export interface BubbleOptions {
    attribute: string;
    element: string;
    id: string[];
    left: number;
    show: boolean;
    type: BubbleType;
    top: number;
    value: string;
}

export interface Actions {
    setXml: (xml: Xml) => void;
    showBubble: (askOptions: Partial<BubbleOptions>) => void;
}

export interface AskerOptions {
    actions: Actions;
    defaultValue: string;
    id: string[];
    parameter?: AskerParameter;
    xml: Xml;
}

export interface Element {
    _?: string;
    $?: {[key: string]: string};
    $$?: Element[];
    '#collapsed'?: boolean;
    '#name': string;
}
