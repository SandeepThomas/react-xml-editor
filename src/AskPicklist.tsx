import * as React from 'react';

import {
    Actions,
    AskerParameter,
    AskPicklistCaptionMenuItem,
    AskPicklistMenuItem,
    Xml,
} from './types';
import { updateNode } from './Util';

interface Props {
    actions: Actions;
    id: string[];
    parameter: AskerParameter;
    xml: Xml;
}

const hasCaption = (menuItem: AskPicklistMenuItem): menuItem is AskPicklistCaptionMenuItem => {
    return (menuItem as AskPicklistCaptionMenuItem).caption !== undefined;
};

export default class AskPicklist extends React.Component<Props> {
    public render(): React.ReactNode {
        const { parameter } = this.props;

        const menuItems = parameter.map((menuItem) => {
            let caption;
            let value: string;
            if (hasCaption(menuItem)) {
                caption = <span className="explainer">{ menuItem.caption }</span>;
                value = menuItem.value;
            } else {
                value = menuItem;
            }

            return (
                <div
                    className="menuItem focusme techno"
                    key={ value }
                    onClick={(e: React.MouseEvent) => {
                        this.onClick(value);
                        e.preventDefault();
                    }}
                >
                    <span className="punc">"</span>
                    { value }
                    <span className="punc">"</span>
                    { caption }
                </div>
            );
        });

        return (
            <div className="menu">
                { menuItems }
            </div>
        );
    }

    private onClick(value: string) {
        const { actions, id, xml } = this.props;
        actions.setXml(updateNode(xml, id, value));
        actions.showBubble({
            show: false,
        });
    }
}
