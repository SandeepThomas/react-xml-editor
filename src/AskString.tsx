import * as React from 'react';

import { IActions, IXml } from './types';
import { updateNode } from './Util';

export enum EAskStringType {
    LONG,
    SHORT,
}

interface IProps {
    actions: IActions;
    defaultValue: string;
    id: string;
    type: EAskStringType;
    xml: IXml;
}

interface IState {
    value: string;
}

export default class AskString extends React.Component<IProps, IState> {
    public constructor(props: IProps) {
        super(props);
        this.onChange = this.onChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.state = {
            value: props.defaultValue,
        };
    }

    public render(): React.ReactNode {
        const { type } = this.props;
        return (
            <form onSubmit={ this.onSubmit }>
                { type === EAskStringType.LONG ? this.getLongString() : this.getShortString() }
            </form>
        );
    }

    private getShortString(): React.ReactNode {
        const { value } = this.state;

        return (
            <>
                <input
                    name="val"
                    className="textbox focusme"
                    onChange={ this.onChange }
                    value={ value }
                />
                <input type="submit" value="OK"/>
            </>
        );
    }

    private getLongString(): React.ReactNode {
        const { value } = this.state;
        return (
            <>
                <textarea
                    name="val"
                    className="textbox focusme"
                    onChange={ this.onChange }
                    value={value}
                 />
                <div className="submitline">
                    <input type="submit" value="OK" />
                </div>
            </>
        );
    }

    private onSubmit(e: React.FormEvent<HTMLFormElement>) {
        const { actions, id, xml } = this.props;
        const { value } = this.state;

        actions.setXml(updateNode(xml, id, value));
        actions.showBubble({
            show: false,
        });
        e.preventDefault();
    }

    private onChange(e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) {
        this.setState({
            value: e.target.value,
        });
    }
}
