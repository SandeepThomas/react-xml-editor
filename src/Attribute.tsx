import * as React from 'react';

import { EBubbleType, IActions, IBubbleOptions } from './types';

interface IProps {
    actions: IActions;
    element: string;
    id: string;
    name: string;
    value: string;
}

export default class Attribute extends React.Component<IProps> {
    private ref: React.RefObject<HTMLSpanElement>;

    public constructor(props: IProps) {
        super(props);
        this.ref = React.createRef();
        this.onClickName = this.onClickName.bind(this);
        this.onClickValue = this.onClickValue.bind(this);
    }

    public render(): React.ReactNode {
        const { name, value } = this.props;

        return (
            <span className="attribute">
                <span className="punc"> </span>
                <span className="name" onClick={ this.onClickName }>{ name }</span>
                <span className="punc">=</span>
                <span className="valueContainer" onClick={ this.onClickValue } ref={ this.ref }>
                    <span className="punc">"</span>
                    <span className="value">{ value }</span>
                    <span className="punc">"</span>
                </span>
            </span>
        );
    }

    private onClickName(): void {
        const { actions, element, id, name, value } = this.props;
        const bubbleOptions: Partial<IBubbleOptions> = {
            attribute: name,
            element,
            id,
            show: true,
            type: EBubbleType.MENU,
            value,
        };
        if (this.ref.current) {
            const rect = this.ref.current.getBoundingClientRect();
            bubbleOptions.left = rect.left;
            bubbleOptions.top = rect.top;
        }
        actions.showBubble(bubbleOptions);
    }

    private onClickValue(): void {
        const { actions, element, id, name, value } = this.props;
        const bubbleOptions: Partial<IBubbleOptions> = {
            attribute: name,
            element,
            id,
            show: true,
            type: EBubbleType.ASKER,
            value,
        };
        if (this.ref.current) {
            const rect = this.ref.current.getBoundingClientRect();
            bubbleOptions.left = rect.left;
            bubbleOptions.top = rect.top;
        }
        actions.showBubble(bubbleOptions);
        // event.stopPropagation();
    }
}
