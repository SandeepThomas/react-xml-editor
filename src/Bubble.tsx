import * as React from 'react';

import { EBubbleType, IActions, IDocSpec, IMenuItemSpec, IXml } from './types';
import { askLongString, updateNode } from './Util';

interface IProps {
    actions: IActions;
    attribute: string;
    docSpec: IDocSpec;
    element: string;
    id: string;
    left: number;
    show: boolean;
    top: number;
    type: EBubbleType;
    value: string;
    xml: IXml;
}

export default class Bubble extends React.Component<IProps> {
    public constructor(props: IProps) {
        super(props);
        this.onChange = this.onChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }

    public render(): React.ReactNode {
        const {
            id,
            show,
        } = this.props;
        if (!show) {
            return null;
        }

        if (id.match(/~\$~[^~]+$/)) {
            return this.getAttributeBubble();
        } else if (id.match(/~\$\$~[0-9]+$/)) {
            return this.getElementBubble();
        } else if (id.match(/~_$/)) {
            return this.getTextBubble();
        }

        return null;
    }

    private getTextBubble(): React.ReactNode {
        const { actions, id, left, top, value, xml } = this.props;
        return (
            <div className="nerd" id="xonomyBubble" style={{ left, top, display: 'block' }}>
                <div className="inside">
                    <div id="xonomyBubbleContent">
                        { askLongString({
                            actions,
                            defaultValue: value,
                            id,
                            xml,
                        }) }
                    </div>
                </div>
            </div>
        );
    }

    private getElementBubble(): React.ReactNode {
        const { actions, docSpec, element, id, left, top, xml } = this.props;

        if (docSpec.elements && docSpec.elements[element] && docSpec.elements[element].menu) {
            const menu = docSpec.elements[element].menu;
            const menuItems = (menu as IMenuItemSpec[]).map((menuItemSpec, index) => (
                <div
                    className="menuItem focusme"
                    key={ index }
                    onClick={ async () => {
                        actions.setXml(await menuItemSpec.action(xml, id, menuItemSpec.actionParameter));
                        actions.showBubble({ show: false });
                    } }
                >
                    { menuItemSpec.caption }
                </div>
            ));
            return (
                <div className="nerd" id="xonomyBubble" style={{ left, top, display: 'block' }}>
                    <div className="inside">
                        <div id="xonomyBubbleContent">
                            <div className="menu">
                                {menuItems}
                            </div>
                        </div>
                    </div>
                </div>
            );
        }
        return null;
    }

    private getAttributeBubble(): React.ReactNode {
        const {
            actions,
            attribute,
            docSpec,
            id,
            element,
            left,
            top,
            type,
            value,
            xml,
        } = this.props;
        if (type === EBubbleType.ASKER) {
            if (docSpec.elements &&
                docSpec.elements[element] &&
                docSpec.elements[element].attributes &&
                docSpec.elements[element].attributes[attribute] &&
                docSpec.elements[element].attributes[attribute].asker) {
                return (
                    <div className="nerd" id="xonomyBubble" style={{ left, top, display: 'block' }}>
                        <div className="inside">
                            <div id="xonomyBubbleContent">
                            { (docSpec.elements[element].attributes[attribute].asker as any)({
                                actions,
                                defaultValue: value,
                                id,
                                xml,
                            }) }
                            </div>
                        </div>
                    </div>
                );
            }
        } else if (type === EBubbleType.MENU) {
            if (docSpec.elements &&
                docSpec.elements[element] &&
                docSpec.elements[element].attributes &&
                docSpec.elements[element].attributes[attribute] &&
                docSpec.elements[element].attributes[attribute].menu) {
                    const menu = docSpec.elements[element].attributes[attribute].menu;
                    const menuItems = (menu as IMenuItemSpec[]).map((menuItemSpec, index) => (
                        <div
                            className="menuItem focusme"
                            key={ index }
                            onClick={ async () => {
                                // setXml(menuItemSpec.action(xml, menuItemSpec.actionParameter));
                                actions.setXml(await menuItemSpec.action(xml, id, menuItemSpec.actionParameter));
                                actions.showBubble({ show: false });
                            } }
                        >
                            { menuItemSpec.caption }
                        </div>
                    ));

                    return (
                        <div className="nerd" id="xonomyBubble" style={{ left, top, display: 'block' }}>
                            <div className="inside">
                                <div id="xonomyBubbleContent">
                                    <div className="menu">
                                        {menuItems}
                                    </div>
                                </div>
                            </div>
                        </div>
                    );
            }
        }
        return null;
    }

    private onSubmit(event: React.FormEvent<HTMLFormElement>): void {
        const { actions, id, value, xml } = this.props;
        actions.setXml(updateNode(xml, id, value));
        actions.showBubble({
            show: false,
        });
        event.preventDefault();
    }

    private onChange(event: React.FormEvent<HTMLInputElement>): void {
        const { actions } = this.props;
        actions.showBubble({
            value: (event.target as HTMLInputElement).value,
        });
    }
}
