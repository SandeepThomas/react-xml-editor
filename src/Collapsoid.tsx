import * as React from 'react';

import { IElement } from './Parser';

interface IProps {
    elements: IElement[];
}

export default class Collapsoid extends React.Component<IProps> {
    private static maxLength: number = 35;

    public render(): React.ReactNode {
        const { elements } = this.props;
        return this.renderElements(elements);
    }

    private renderElements(elements: IElement[]): string {
        let collapsoid = '';
        elements.every((element: IElement) => {
            if (element._ && element['#name'] === '__text__') {
                collapsoid += element._ + ' ';
            }
            if (element.$$) {
                collapsoid += this.renderElements(element.$$) + ' ';
            }
            return collapsoid.length < Collapsoid.maxLength;
        });
        return collapsoid.trim().slice(0, Collapsoid.maxLength);
    }
}
