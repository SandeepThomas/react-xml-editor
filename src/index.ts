export { default as XmlEditor } from './XmlEditor';
