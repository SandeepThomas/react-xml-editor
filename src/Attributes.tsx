import * as React from 'react';

import Attribute from './Attribute';
import { Actions } from './types';
import { push } from './Util';

interface Props {
    actions: Actions;
    attributes: {[key: string]: string};
    id: string[];
    element: string;
}

export default class Attributes extends React.Component<Props> {
    public render(): React.ReactNode {
        const { actions, attributes, element, id } = this.props;
        return (
            <span className="attributes">
                { Object.keys(attributes).map((name: string) => (
                    <Attribute
                        actions={ actions }
                        element={ element }
                        id={ push(id, '$', name) }
                        key={ name }
                        name={ name }
                        value={ this.props.attributes[name] }
                    />
                )) }
            </span>
        );
    }
}
