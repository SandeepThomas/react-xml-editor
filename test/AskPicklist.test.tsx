import {describe, expect, jest, test} from '@jest/globals';
import renderer from 'react-test-renderer';
import * as React from 'react';

import AskPicklist from '../src/AskPicklist';
import { Actions } from '../src/types';

describe('AskPicklist component', () => {

    test('AskPicklist renders properly', () => {
        const parameter = [{
            value: 'short', caption: 'short'
        },{
            value: 'medium', caption: 'medium',
        }, 'long'];

        const setXml = jest.fn();
        const showBubble = jest.fn();
        const preventDefault = jest.fn();

        const component = renderer.create(
            <AskPicklist
                actions={{setXml, showBubble} as Actions}
                id={['id']}
                parameter={parameter}
                xml={{}}
            />
        );
        const tree = component.toJSON();
        tree.children[0].props.onClick({
            preventDefault,
        });
        expect(tree).toMatchSnapshot();
        expect(setXml.mock.calls[0][0]).toEqual({
            id: 'short',
        });
        expect(showBubble.mock.calls[0][0]).toEqual({
            show: false,
        });
        expect(preventDefault.mock.calls.length).toBe(1);
    });

});