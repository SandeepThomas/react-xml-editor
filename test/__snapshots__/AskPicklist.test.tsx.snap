// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AskPicklist component AskPicklist renders properly 1`] = `
<div
  className="menu"
>
  <div
    className="menuItem focusme techno"
    onClick={[Function]}
  >
    <span
      className="punc"
    >
      "
    </span>
    short
    <span
      className="punc"
    >
      "
    </span>
    <span
      className="explainer"
    >
      short
    </span>
  </div>
  <div
    className="menuItem focusme techno"
    onClick={[Function]}
  >
    <span
      className="punc"
    >
      "
    </span>
    medium
    <span
      className="punc"
    >
      "
    </span>
    <span
      className="explainer"
    >
      medium
    </span>
  </div>
  <div
    className="menuItem focusme techno"
    onClick={[Function]}
  >
    <span
      className="punc"
    >
      "
    </span>
    long
    <span
      className="punc"
    >
      "
    </span>
  </div>
</div>
`;
